# 1d Convolutional Neural Networks

In [0]:

# Imports
import numpy as np
import os
# from utils.utilities import *
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
import tensorflow as tf
import pandas as pd
import tensorflow as tf
from tensorflow import keras

### About the Data

Data is from people doing one of these activities (WALKING, WALKING_UPSTAIRS, WALKING_DOWNSTAIRS, SITTING, STANDING, LAYING) <br><br>
The data is from a smartphone and provides the body acceleration, general acceleration and a gyroscope reading <br><br>
Data is from: https://archive.ics.uci.edu/ml/datasets/human+activity+recognition+using+smartphones 

### Read in data from functions provided below

In [0]:
X_train, labels_train, list_ch_train = read_data_train() # train
X_test, labels_test, list_ch_test = read_data_test()



In [74]:
print(X_train.shape)

(7352, 128, 9)



#### Explaining the shape of the data 
There are 7,352 data points that are classified as doing one of those activities <br>
There are 128 time steps <br>
There are 9 values for each time step for each data point for the x,y,z values of the body 
acceleration, general acceleration and gyrscope reading. <br>

### Use train_test_split to create training and validation data and labels

In [0]:
# your code here
from sklearn.model_selection import train_test_split

One hot coding is when each row represents one label and it the nonzero column represents what the label is<br>
 [0,0,0,0,0,1] <br>
           ^ This label is LAYING, which was previously represented with a 5
           <br><br>

### One hot encoding functions are provided at the bottom of this notebook <br> Use these functions to convert the training, validation and testing labels

In [0]:
# your code here
labels_train = one_hot(labels_train)
labels_test = one_hot(labels_test)
X_train, X_val, labels_train, labels_val = train_test_split(X_train, labels_train, stratify = labels_train)

### Define your Keras model <br>
You wlil want to use

See for examples on creating Keras modelshttps://github.com/keras-team/keras/blob/master/examples/mnist_cnn.py

In [77]:
# fill in with your code below
model = keras.Sequential()
model.add(keras.layers.Conv1D(filters = 10, kernel_size = 2, padding = 'same', activation = tf.nn.relu,input_shape = (128,9)))
model.add(keras.layers.GlobalMaxPooling1D())
model.add(keras.layers.Dense(6,activation=tf.nn.sigmoid))
print(model.summary())  # view model
    
    


_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv1d_11 (Conv1D)           (None, 128, 10)           190       
_________________________________________________________________
global_max_pooling1d_3 (Glob (None, 10)                0         
_________________________________________________________________
dense_7 (Dense)              (None, 6)                 66        
Total params: 256
Trainable params: 256
Non-trainable params: 0
_________________________________________________________________
None


### Compile your model

In [0]:
lr =  0.0008# choose a learning rate, this will be a good paramter to tune
model.compile( optimizer=tf.train.AdamOptimizer(lr), loss=keras.losses.categorical_crossentropy, metrics = ['accuracy'] )

### Fit your model

In [80]:
history = model.fit(X_train,
                    labels_train,
                    epochs= 20,
                    batch_size= 100,
                    validation_data=(X_val, labels_val),
                    verbose=1)

Train on 5514 samples, validate on 1838 samples
Epoch 1/20
5514/5514 [==============================] - 1s 182us/step - loss: 1.6586 - acc: 0.2106 - val_loss: 1.6154 - val_acc: 0.2818
Epoch 2/20
5514/5514 [==============================] - 1s 188us/step - loss: 1.5685 - acc: 0.3874 - val_loss: 1.5195 - val_acc: 0.5038
Epoch 3/20
5514/5514 [==============================] - 1s 193us/step - loss: 1.4635 - acc: 0.5832 - val_loss: 1.4071 - val_acc: 0.6181
Epoch 4/20
5514/5514 [==============================] - 1s 194us/step - loss: 1.3486 - acc: 0.6848 - val_loss: 1.2919 - val_acc: 0.7405
Epoch 5/20
5514/5514 [==============================] - 1s 194us/step - loss: 1.2325 - acc: 0.7748 - val_loss: 1.1741 - val_acc: 0.8107
Epoch 6/20
5514/5514 [==============================] - 1s 194us/step - loss: 1.1131 - acc: 0.8179 - val_loss: 1.0566 - val_acc: 0.8308
Epoch 7/20
5514/5514 [==============================] - 1s 194us/step - loss: 0.9973 - acc: 0.8393 - val_loss: 0.9444 - val_acc: 0.8498


5514/5514 [==============================] - 1s 196us/step - loss: 0.8900 - acc: 0.8531 - val_loss: 0.8441 - val_acc: 0.8553
Epoch 9/20
5514/5514 [==============================] - 1s 195us/step - loss: 0.7951 - acc: 0.8616 - val_loss: 0.7571 - val_acc: 0.8700
Epoch 10/20
5514/5514 [==============================] - 1s 193us/step - loss: 0.7121 - acc: 0.8723 - val_loss: 0.6827 - val_acc: 0.8781
Epoch 11/20
5514/5514 [==============================] - 1s 193us/step - loss: 0.6434 - acc: 0.8763 - val_loss: 0.6211 - val_acc: 0.8814
Epoch 12/20
5514/5514 [==============================] - 1s 188us/step - loss: 0.5865 - acc: 0.8901 - val_loss: 0.5699 - val_acc: 0.8868
Epoch 13/20
5514/5514 [==============================] - 1s 190us/step - loss: 0.5402 - acc: 0.8919 - val_loss: 0.5271 - val_acc: 0.8874
Epoch 14/20
5514/5514 [==============================] - 1s 191us/step - loss: 0.5015 - acc: 0.8945 - val_loss: 0.4919 - val_acc: 0.8912
Epoch 15/20
5514/5514 [==============================]

Epoch 16/20
5514/5514 [==============================] - 1s 189us/step - loss: 0.4406 - acc: 0.9042 - val_loss: 0.4359 - val_acc: 0.8977
Epoch 17/20
5514/5514 [==============================] - 1s 188us/step - loss: 0.4168 - acc: 0.9095 - val_loss: 0.4143 - val_acc: 0.9010
Epoch 18/20
5514/5514 [==============================] - 1s 182us/step - loss: 0.3948 - acc: 0.9128 - val_loss: 0.3935 - val_acc: 0.9015
Epoch 19/20
5514/5514 [==============================] - 1s 181us/step - loss: 0.3764 - acc: 0.9160 - val_loss: 0.3765 - val_acc: 0.9042
Epoch 20/20
5514/5514 [==============================] - 1s 179us/step - loss: 0.3598 - acc: 0.9169 - val_loss: 0.3606 - val_acc: 0.9010


### Code below prints the test loss and accuracy <br>
### Before you print your accuracy and loss, play around with the parameters!!

In [81]:
score = model.evaluate(X_test, labels_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

('Test loss:', 0.5219826167231298)
('Test accuracy:', 0.829657278588395)


### Provided Functions Below

In [0]:
def read_data_train():
  """ Read data """

  # Fixed params
  n_class = 6
  n_steps = 128

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/y_train.txt'
  labels = pd.read_csv(label_path, header = None)

  list_of_channels = ['body_acc_x', 'body_acc_y', 'body_acc_z', 'body_gyro_x',
  'body_gyro_y', 'body_gyro_z', 'total_acc_x', 'total_acc_y', 'total_acc_z']

  X = np.zeros((len(labels), n_steps, len(list_of_channels)))

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_acc_x_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,0] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_acc_y_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,1] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_acc_z_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,2] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_gyro_x_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,3] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_gyro_y_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,4] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_gyro_z_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,5] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/total_acc_x_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,6] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/total_acc_y_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,7] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/total_acc_z_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,8] = dat_.as_matrix()


  # Return 
  return X, labels[0].values, list_of_channels

In [0]:
def read_data_test():
  """ Read data """

  # Fixed params
  n_class = 6
  n_steps = 128

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/y_test.txt'
  labels = pd.read_csv(label_path, header = None)

  list_of_channels = ['body_acc_x', 'body_acc_y', 'body_acc_z', 'body_gyro_x',
  'body_gyro_y', 'body_gyro_z', 'total_acc_x', 'total_acc_y', 'total_acc_z']

  X = np.zeros((len(labels), n_steps, len(list_of_channels)))

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_acc_x_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,0] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_acc_y_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,1] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_acc_z_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,2] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_gyro_x_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,3] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_gyro_y_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,4] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_gyro_z_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,5] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/total_acc_x_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,6] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/total_acc_y_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,7] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/total_acc_z_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,8] = dat_.as_matrix()


  # Return 
  return X, labels[0].values, list_of_channels

In [0]:
def one_hot(labels, n_class = 6):
	""" One-hot encoding """
	expansion = np.eye(n_class)
	y = expansion[:, labels-1].T
	assert y.shape[1] == n_class, "Wrong number of labels!"

	return y

In [0]:
def get_batches(X, y, batch_size = 100):
	""" Return a generator for batches """
	n_batches = len(X) // batch_size
	X, y = X[:n_batches*batch_size], y[:n_batches*batch_size]

	# Loop over batches and yield
	for b in range(0, len(X), batch_size):
		yield X[b:b+batch_size], y[b:b+batch_size]